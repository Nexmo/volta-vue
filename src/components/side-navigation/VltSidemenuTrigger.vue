<template>
  <p
    v-click-outside
    :id="id"
    class="Vlt-sidemenu__trigger"
    :class="classObject"
    @click="isActive = !isActive">
    <vlt-icon :icon="icon" :color="color" />
    <span class="Vlt-sidemenu__label">{{label}}</span>
    <vlt-badge v-if="badge" :color="badgeColor">{{badge}}</vlt-badge>
  </p>
</template>

<script>
/* eslint-disable no-param-reassign */
import Vue from 'vue';
import VltIcon from '../VltIcon';
import VltBadge from '../VltBadge';

Vue.directive('click-outside', {
  bind(el, binding, vnode) {
    vnode.context.event = () => {
      if (vnode.context.collapsed) {
        vnode.context.isActive = false;
      }
    };

    vnode.context.stopProp = (event) => {
      event.stopPropagation();
    };

    el.addEventListener('click', vnode.context.stopProp);
    document.body.addEventListener('click', vnode.context.event);
  },

  unbind(el, binding, vnode) {
    el.removeEventListener('click', vnode.context.stopProp);
    document.body.removeEventListener('click', vnode.context.event);
  },
});

export default {
  name: 'vlt-sidemenu-trigger',

  props: {
    active: Boolean,
    badge: String,
    badgeColor: String,
    color: String,
    collapsed: Boolean,
    icon: String,
    id: String,
    label: String,
  },

  data() {
    return {
      isActive: this.active,
    };
  },

  computed: {
    classObject() {
      return {
        'Vlt-sidemenu__trigger_active': this.isActive,
        'Vlt-sidemenu__trigger_current': this.isActive,
      };
    },
  },

  components: {
    VltBadge,
    VltIcon,
  },

  watch: {
    active(value) {
      this.isActive = value;
    },
  },
};
</script>

<style lang="scss" scoped>
  .Vlt-badge {
    margin-right: 20px !important
  }
</style>
