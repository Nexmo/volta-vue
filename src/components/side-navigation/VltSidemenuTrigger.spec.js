import { shallowMount } from '@vue/test-utils';
import VltSidemenuTrigger from './VltSidemenuTrigger';

describe('vlt-sidemenu-trigger', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(VltSidemenuTrigger, {
      propsData: {
        icon: 'rocket',
      },
    });
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test('renders correctly with a rocket icon', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
