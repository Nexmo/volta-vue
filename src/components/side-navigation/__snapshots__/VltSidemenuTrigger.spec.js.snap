// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vlt-sidemenu-trigger renders correctly with a rocket icon 1`] = `
<p
  class="Vlt-sidemenu__trigger"
>
  <vlt-icon-stub
    clickable="true"
    color=""
    icon="rocket"
    noclass="false"
    pointer="false"
    small="false"
    smaller="false"
  />
  <span
    class="Vlt-sidemenu__label"
  />
  <!--v-if-->
</p>
`;
